Update references to 1.2 spec, re-ordering, content updates (#9)
ResearchObject/packaging_data_with_ro-crate: episodes/14-common-pitfalls.md

Reviewer: Really like this new page, thanks for putting it together!

::::::::::::::::::::::::::::::::::::::::::::::::::

## Introduction

Creating RO-Crate packages by hand is a great way to understand their structure, but it’s easy to run into small issues that can break the crate or make it difficult to interpret. This section outlines the most frequent problems and how to avoid them.

---

## Common Mistakes

### 1. Invalid or malformed JSON

- **Problem**: Forgetting commas, quotes, or brackets.
- **Fix**: Use an editor with JSON syntax highlighting (e.g., VS Code) and test your crate in the [JSON-LD Playground](https://json-ld.org/playground/).

> Tip: Run your `ro-crate-metadata.json` through a JSON validator before uploading it anywhere.

---

### 2. Missing `@context` or `@graph`

- **Problem**: Forgetting to include the required top-level keys in the RO-Crate metadata file.
- **Fix**: Make sure every `ro-crate-metadata.json` starts like this:

```json
{
  "@context": "https://w3id.org/ro/crate/1.2/context",
  "@graph": [ /* your entities here */ ]
}
```

### 3. Using nested JSON instead of flat structure

- **Problem**: Embedding full descriptions of related entities instead of referencing them with `@id`.

> - **Fix**: Use cross-referencing. Define each entity once and refer to it using its `@id`.

Reviewer: ```suggestion
- **Fix**: Use cross-referencing. Define each entity once in the `@graph` array and refer to it using its `@id`.
```

### 4. Repeating or mismatched @id values

> - **Problem**: Accidentally defining two entities with the same @id, or referencing an entity with an @id that doesn’t exist.

Reviewer: ```suggestion
- **Problem**: Accidentally defining two entities with the same @id, or referencing an entity with an @id that doesn’t exist in the `@graph`.
```

> - **Fix**: Ensure every @id is unique and points to an entity defined in the metadata.

Reviewer: ```suggestion
- **Fix**: Ensure every @id is unique and cross-references point to an entity defined in the metadata.
```

### 5. Forgetting required metadata on key entities

> - **Problem**: The root dataset doesn’t have name or license, or files are missing type.

Reviewer: ```suggestion
- **Problem**: The root dataset doesn’t have `name` or `license`, or files are missing `@type`.
```

- **Fix**: Always include the required fields for the root (./) and for files:
```json
{
  "@id": "./",
  "@type": "Dataset",
  "name": "My Crate",
  "description": "An example research dataset.",
  "license": { "@id": "http://spdx.org/licenses/CC0-1.0" }
}
```
:::::::::::::::::::::::::::::::::::::::: challenge
## Spot the errors
Here's an incorrect metadata example. Identify and fix the issues.

```json
{
  "@context": "https://w3id.org/ro/crate/1.2/context",
  "@graph": [
    {
      "@id": "./",
      "@type": "Folder",
      "name": "Rainfall data",
      "hasPart": {
        "@id": "data.csv",
        "name": "Rainfall Katoomba",
        "encodingFormat": "comma separated values",
      }
    }
  ]
}
```
:::::::::::::::  solution
1. The `@type` for the root should be "Dataset", not "Folder". 
2. The hasPart value should reference an @id, not contain the full entity.
3. The data.csv entity should be defined separately in the @graph.
4. The encodingFormat should be a valid IANA media type (e.g., "text/csv").
5. Missing required properties (description, datePublished, license) on the root dataset.

> ::::::::::::::::::::::::::::::::::::::::

Reviewer: think you need an extra `:` row here to close the challenge block - might be the source of the build error (but haven't tested locally)

```suggestion
::::::::::::::::::::::::::::::::::::::::
::::::::::::::::::::::::::::::::::::::::
```

Author: Thank you! Not sure why I didn't see that but it looks like it's fixed now.

> 

Reviewer: Suggesting a short extra section to highlight that these checks can be done automatically (more for self learning than for live teaching, but could demo if there is time). I am not sure if the Playground catches all of the problems mentioned, but I think the CRS4 validator does.

`````suggestion

## RO-Crate validation tools

The [RO-Crate Playground](https://ro-crate.ldaca.edu.au) and the [rocrate-validator](https://rocrate-validator.readthedocs.io/en/latest/) Python package can both be used to check for errors in your RO-Crate metadata. Each of them performs a different set of checks.

:::::::::::::::::::::::::::::::::::::::: challenge

Copy and paste your completed JSON-LD into the [RO-Crate Playground](https://ro-crate.ldaca.edu.au), then click "Validate". Are there any issues?

::::::::::::::::::::::::::::::::::::::::

:::::::::::::::::::::::::::::::::::::::: challenge

Install the [rocrate-validator](https://rocrate-validator.readthedocs.io/en/latest/) and run it on the command line against your completed RO-Crate:

```
rocrate-validator validate <path_to_rocrate>
```

Are there any issues? Are the results different to the RO-Crate Playground?

::::::::::::::::::::::::::::::::::::::::

`````
